### PR TITLE
support `withincode pointcut` in the event definition

### PR DESCRIPTION
**Pull request by yzhang90**

### CODE REVIEW

File: src/main/java/javamop/parser/ast/visitor/GenericVisitor.java [hunk 1]
```diff
@@ -107,6 +107,8 @@
 
 	public R visit(WithinPointCut p, A arg);
 
+    public R visit(WithincodePointCut p, A arg);
```
Comment (owolabileg) [inline src/main/java/javamop/parser/ast/visitor/GenericVisitor.java:110]: Minor: One more horizontal space

File: src/main/java/javamop/parser/ast/visitor/MoveNotInwardPointcutVisitor.java [hunk 2]
```diff
@@ -108,6 +89,10 @@ public PointCut visit(WithinPointCut p, Object arg) {
 		return p;
 	}
 
+    public PointCut visit(WithincodePointCut p, Object arg) {
+        return p;
+    }
+
```
Comment (owolabileg) [inline src/main/java/javamop/parser/ast/visitor/MoveNotInwardPointcutVisitor.java:95]: One more horizontal space

**Comment by owolabileg** [comment]
1. What's the use case (e.g., an example spec) that requires a withincode pointcut?
2. There are several other cases where more horizontal space should be added to new code, for consistency. I pointed out a couple.
3. What's the maximum number of imports from the same package before we import `*` from that package? I see several cases where the new changes imports `*` for relatively few imports. Perhaps keep the separate imports?

**Comment by yzhang90** [comment]
> 1. What's the use case (e.g., an example spec) that requires a withincode pointcut?
The withincode is used to pinpoint the exact instrumentation location. For example, I only want to instrument before db.put method when it is called inside dbStorage.put method.
```
    event putdbLock before ():
        call(* HaloDB.put(byte[], byte[]))
        && withincode(* HaloDBStorageEngine.put(byte[], byte[], int)) {
            System.out.println("Before acquire the putLock");
            dbPutLock.lock();
            System.out.println("After acquire the putLock");
    }
```
in this event we want to synchronize the db.put method and monitor state update code in the  HaloDBStorageEngine.put(..) function.

> 4. Please check that these are all the places that need to be changed, e.g., is withincode in a raw spec parsed properly?
I tried this change on a raw spec.

**Comment by yzhang90** [comment]
> 3. What's the maximum number of imports from the same package before we import `*` from that package? I see several cases where the new changes imports `*` for relatively few imports. Perhaps keep the separate imports?
I change them to imports `*` because in the visitor class, we basically import all the class from the ast package. Do we need to list them one by one in the header?

**Comment by msaxena2** [comment]
@yzhang90 Looks good, apart from the formatting issue which has already been mentioned - 
1. Seems like there is a file for unit testing the parser, although the number of tests seems to be extremely limited. Perhaps we can start improving the testing situation by adding a test for withincode?
2. Also extend the integration test using a simple spec that exercises withincode? This perhaps lies beyond the scope of this PR, since the integration testing may require a little more setup.

**Pull request merged**